middleware-stack test file compile fails in typescript 3.2

Because typescript 3.2 introduces the new feature of [type checking](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#caveats) in `bind`, `call`, and `apply`, [typing issues](https://github.com/aws/aws-sdk-js-v3/blob/master/packages/middleware-stack/src/index.spec.ts#L36~L47) not caught before are revealed. 

**issue:**
> @aws-sdk/middleware-stack@0.1.0-preview.1 pretest /Users/zheallan/workspace/npmProj/v3/aws-sdk-js-v3/packages/middleware-stack
> tsc -p tsconfig.test.json

> src/index.spec.ts(51,23): error TS2345: Argument of type '((next: Handler<string[], object>) => Handler<string[], object>) | { priority: number; } | { step: string; }' is not assignable to parameter of type 'FinalizeMiddleware<string[], object, Uint8Array>'.
  Type '(next: Handler<string[], object>) => Handler<string[], object>' is not assignable to type 'FinalizeMiddleware<string[], object, Uint8Array>'.
    Types of parameters 'next' and 'next' are incompatible.
      Types of parameters 'args' and 'args' are incompatible.
        Property 'request' is missing in type 'HandlerArguments<string[]>' but required in type 'FinalizeHandlerArguments<string[], Uint8Array>'.
src/index.spec.ts(96,25): error TS2345: Argument of type '(next: Handler<string[], object>) => Handler<string[], object>' is not assignable to parameter of type 'FinalizeMiddleware<string[], object, Uint8Array>'.
src/index.spec.ts(98,13): error TS2345: Argument of type '(next: Handler<string[], object>) => Handler<string[], object>' is not assignable to parameter of type 'FinalizeMiddleware<string[], object, Uint8Array>'.

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,8 @@
+language: node_js
+node_js:
+  - "10.9.0"
+install:
+  - npm install
+  - npm run bootstrap
+script:
+  - npm run test

--- a/packages/middleware-stack/src/index.spec.ts
+++ b/packages/middleware-stack/src/index.spec.ts
@@ -3,21 +3,27 @@ import {
     Handler,
     HandlerArguments,
     FinalizeHandlerArguments,
+    Middleware,
+    HandlerExecutionContext,
+    FinalizeMiddleware,
+    FinalizeHandler
 } from "@aws-sdk/types";
 
 type input = Array<string>;
 type output = object;
 
-function concatMiddleware(
-    message: string,
-    next: Handler<input, output>
-): Handler<input, output> {
-    return (args: HandlerArguments<input>): Promise<output> => next({
-        ...args,
-        input: args.input.concat(message),
-    });
+//return tagged union to make compiler happy
+function getConcatMiddleware(message: string): Middleware<input, output> | FinalizeMiddleware<input, output> {
+    return (
+        next: Handler<input, output>,
+    ): Handler<input, output> => {
+        return (args: HandlerArguments<input>): Promise<output> => next({
+            ...args,
+            input: args.input.concat(message),
+        });
+    }
+    
 }
-
 function shuffle<T>(arr: Array<T>): Array<T> {
     arr = [...arr];
     for (let i = arr.length; i > 0; i--) {
@@ -33,22 +39,22 @@ describe('MiddlewareStack', () => {
         const stack = new MiddlewareStack<input, output>();
 
         const middleware = shuffle([
-            [concatMiddleware.bind(null, 'second')],
-            [concatMiddleware.bind(null, 'first'), {priority: 10}],
-            [concatMiddleware.bind(null, 'fourth'), {step: 'build'}],
+            [getConcatMiddleware('second'), {}],
+            [getConcatMiddleware('first'), {priority: 10}],
+            [getConcatMiddleware('fourth'), {step: 'build'}],
             [
-                concatMiddleware.bind(null, 'third'),
+                getConcatMiddleware('third'),
                 {step: 'build', priority: 1}
             ],
-            [concatMiddleware.bind(null, 'fifth'), {step: 'finalize'}],
+            [getConcatMiddleware('fifth'), {step: 'finalize'}],
             [
-                concatMiddleware.bind(null, 'sixth'),
+                getConcatMiddleware('sixth'),
                 {step: 'finalize', priority: -1}
             ],
         ]);
 
         for (const [mw, options] of middleware) {
-            stack.add(mw, options);
+            stack.add(mw as any, options);
         }
 
         const inner = jest.fn(({input}: FinalizeHandlerArguments<input>) => {
@@ -71,8 +77,11 @@ describe('MiddlewareStack', () => {
 
     it('should allow cloning', async () => {
         const stack = new MiddlewareStack<input, output>();
-        stack.add(concatMiddleware.bind(null, 'second'));
-        stack.add(concatMiddleware.bind(null, 'first'), {priority: 100});
+        stack.add(getConcatMiddleware('second') as Middleware<input, output>);
+        stack.add(
+            getConcatMiddleware('first')  as Middleware<input, output>,
+            {priority: 100}
+        );
 
         const secondStack = stack.clone();
 
@@ -89,13 +98,19 @@ describe('MiddlewareStack', () => {
 
     it('should allow combining stacks', async () => {
         const stack = new MiddlewareStack<input, output>();
-        stack.add(concatMiddleware.bind(null, 'second'));
-        stack.add(concatMiddleware.bind(null, 'first'), {priority: 100});
+        stack.add(getConcatMiddleware('second') as Middleware<input, output>);
+        stack.add(
+            getConcatMiddleware('first') as Middleware<input, output>,
+            {priority: 100}
+        );
 
         const secondStack = new MiddlewareStack<input, output>();
-        secondStack.add(concatMiddleware.bind(null, 'fourth'), {step: 'build'});
         secondStack.add(
-            concatMiddleware.bind(null, 'third'),
+            getConcatMiddleware('fourth') as FinalizeMiddleware<input, output>,
+            {step: 'build'}
+        );
+        secondStack.add(
+            getConcatMiddleware('third') as FinalizeMiddleware<input, output>,
             {step: 'build', priority: 100}
         );
 
@@ -114,10 +129,10 @@ describe('MiddlewareStack', () => {
     });
 
     it('should allow the removal of middleware by constructor identity', async () => {
-        const MyMiddleware = concatMiddleware.bind(null, 'remove me!');
+        const MyMiddleware = getConcatMiddleware('remove me!') as Middleware<input, output>;
         const stack = new MiddlewareStack<input, output>();
         stack.add(MyMiddleware);
-        stack.add(concatMiddleware.bind(null, "don't remove me"));
+        stack.add(getConcatMiddleware("don't remove me") as Middleware<input, output>);
 
         await stack.resolve(
                 ({input}: FinalizeHandlerArguments<Array<string>>) => {
@@ -144,11 +159,11 @@ describe('MiddlewareStack', () => {
     it('should allow the removal of middleware by tag', async () => {
         const stack = new MiddlewareStack<input, output>();
         stack.add(
-            concatMiddleware.bind(null, 'not removed'),
+            getConcatMiddleware('not removed') as Middleware<input, output>,
             {tags: {foo: true, bar: true}}
         );
         stack.add(
-            concatMiddleware.bind(null, 'remove me!'),
+            getConcatMiddleware('remove me!') as Middleware<input, output>,
             {tags: {foo: true, bar: true, baz: true}}
         );
 

--- a/packages/node-http-handler/src/node-http-handler.spec.ts
+++ b/packages/node-http-handler/src/node-http-handler.spec.ts
@@ -27,8 +27,8 @@ afterEach(() => {
 });
 
 describe('NodeHttpHandler', () => {
-    let mockHttpServer: HttpServer = createMockHttpServer().listen(5432);
-    let mockHttpsServer: HttpsServer = createMockHttpsServer().listen(5433);
+    let mockHttpServer: HttpServer = createMockHttpServer().listen(54321);
+    let mockHttpsServer: HttpsServer = createMockHttpsServer().listen(54322);
 
     afterEach(() => {
         mockHttpServer.removeAllListeners('request');

--- a/packages/service-model/src/TreeModel/deepCopy.fixture.ts
+++ b/packages/service-model/src/TreeModel/deepCopy.fixture.ts
@@ -9,10 +9,10 @@ export function deepCopy<T>(arg: T): T {
     }
 
     if (typeof arg === 'object') {
-        return (Object.keys(arg) as Array<keyof T>).reduce((
+        return (Object.keys(arg) as Array<keyof T>).reduce<T>((
             carry: T,
             item: keyof T
-        ) => ({...carry as any, [item]: deepCopy(arg[item])}), {});
+        ) => ({...carry as any, [item]: deepCopy(arg[item])}), <T>{});
     }
 
     return arg;

